### PR TITLE
Adds photo upload description info per bug 38282 and bug 38286

### PR DESCRIPTION
**Pull request by awjrichards-zz**
- Modifies python script to capture full campaign data rather than limited subset
- Python script now stores campaign data in a dict before dumping to json
- Updates campaigns-data.js to latest full campaign data
- Campaign data structure now made up of keyed objects, campaign name is key
- Campaign sorting by alpha now happens in the template
- upload.js now required in app.js
- updated formatUploadDescription() to specs in bug 38282
- Adds specific categories per bug 38286, with the exception of UA, which
  will be dealt with later.

Note that this is a replacement for previous commmits that I've abandoned
due to some serious git fu fail on my part.

### CODE REVIEW

**Comment by awjrichards-zz** [comment]
See sample upload at http://test.wikipedia.org/w/index.php?title=File:Marist_College_Historic_District_1342124590972.jpeg&action=edit

File: assets/www/index.html [hunk 1]
```diff
@@ -16,7 +16,7 @@
   </head>
   <body class='hidden'>
 	<script type="text/html" id="country-list-template">
-		<% _.each(countries, function( country ) { %>
+		<% _.each( _.sortBy( countries, function(country) { return country.desc }), function( country ) { %>
```
Comment (jdlrobson) [inline assets/www/index.html]: whitespace needed after .desc }

File: assets/www/js/app.js [hunk 2]
```diff
@@ -223,10 +223,14 @@ require( [ 'jquery', 'l10n', 'geo', 'api', 'templates', 'monuments', 'preference
 
 	function showPhotoConfirmation(fileUrl) {
 		var comment = 'Uploaded via WLM Mobile App';
-		var text = 'Testing WLM';
 		var uploadConfirmTemplate = templates.getTemplate('upload-confirm-template');
+
```
Comment (jdlrobson) [inline assets/www/js/app.js]: unrelated whitespace

File: assets/www/js/app.js [hunk 3]
```diff
@@ -223,10 +223,14 @@ require( [ 'jquery', 'l10n', 'geo', 'api', 'templates', 'monuments', 'preference
...
 		var fileName = curMonument.generateFilename();
 		console.log("Filename is " + fileName);
+
+		var text = formatUploadDescription( curMonument, CAMPAIGNS[curMonument.country].config, api.userName );
```
Comment (jdlrobson) [inline assets/www/js/app.js]: whitespace around curMonument

File: assets/www/js/upload.js [hunk 4]
```diff
@@ -46,39 +46,49 @@ function dateYMD() {
...
+			'Mobile upload', 
+			'Uploaded with Android WLM App'
+		],
+		cats = campaignConfig.defaultCategories
+			.concat( campaignConfig.autoCategories )
+			.concat( ourCategories ),
+		autoWikiText = campaignConfig.autoWikiText;
+
 	var desc = '';
 	desc += '=={{int:filedesc}}==\n';
```
Comment (jdlrobson) [inline assets/www/js/upload.js]: can we do this in a more generic way? e.g. make sure of templates - it would be good to allow easy changes to this

File: assets/www/js/upload.js [hunk 5]
```diff
@@ -46,39 +46,49 @@ function dateYMD() {
...
-	
-	
+		idField = idTemplate.replace( '$1', monument.id ),
+		license = campaignConfig.defaultOwnWorkLicence, // note the typo in the API field
+		ourCategories = [ 
+			'Mobile upload', 
+			'Uploaded with Android WLM App'
+		],
+		cats = campaignConfig.defaultCategories
+			.concat( campaignConfig.autoCategories )
```
Comment (jdlrobson) [inline assets/www/js/upload.js]: we usually trail the dot '.' elsewhere in the app e.g.

```
cats = campaignConfig.defaultCategories.
concat
```

File: scripts/generate_campaigns.py [hunk 6]
```diff
@@ -14,20 +14,23 @@
...
 
 for camp in campdata:
 	if camp['name'].startswith('wlm'):
-		campaign = {
-				'name': camp['name'].replace('wlm-', ''),
-				'categories': camp['config']['defaultCategories'],
-				'licenses': camp['config']['licensesOwnWork']
-				}
-		if campaign['name'] in namesdata:
+		camp['name'] = camp['name'].replace('wlm-','')
```
Comment (jdlrobson) [inline scripts/generate_campaigns.py:21]: use more convenient dot notation here ( camp.name ) - only use bracketed quotes when  you need to (e.g. variable/reserved keyword/has a - or other keyword character)

File: scripts/generate_campaigns.py [hunk 7]
```diff
@@ -14,20 +14,23 @@
...
+		if camp['name'] in namesdata:
+			campaign = camp
 			campaign['desc'] = namesdata[campaign['name']]
-			# Only pickup campaings that have a description set
-			campaigns.append(campaign)
+			campaigns.append( campaign )
```
Comment (jdlrobson) [inline scripts/generate_campaigns.py:25]: unrelated change

**Comment by jdlrobson** [comment]
Could we separate this into 2 commits? One for changing the format of CAMPAIGN and one for the description - As Yuvi will testify I like smaller commit for when I need to debug :).

**Comment by jdlrobson** [comment]
Works for me. On test wiki the templates do not get rendered - but I assume that is because test wiki is not setup.

File: scripts/generate_campaigns.py (hunk 6 above)
Comment (awjrichards-zz) [inline scripts/generate_campaigns.py:21]: dude, this is a python script, not js :)

**Comment by awjrichards-zz** [comment]
I split the commit into two commits and addressed most of Jon's issues with the exception of using templates for the photo description text.

**Comment by awjrichards-zz** [comment]
The main reason I haven't implemented the upload description as a template is because I haven't worked with js/templates much before, and certainly not in this kind of context. If you or someone could give me a little guidance, I'd be happy to implement it.

File: scripts/generate_campaigns.py (hunk 6 above)
Comment (jdlrobson) [inline scripts/generate_campaigns.py:21]: Lolz. Fail.

**Comment by awjrichards-zz** [comment]
Ok - I did some reading on underscore and tried implementing a template instead of concat'ing 'desc'. It does not appear to be working, however. You can see the diff of my changes here:

https://gist.github.com/3102170

After rebuilding the app with these changes and trying to upload a photo, I can take a photo or select one from the device, but then I am taken back to the 'Add a photo' screen. logcat shows:
`D/CordovaLog( 8621): Error in success callback: Camera4 = ReferenceError: formatUploadDescription is not defined`
`D/CordovaLog( 8621): file:///android_asset/www/android/cordova-1.8.1.js: Line 254 : Error in success callback: Camera4 = ReferenceError: formatUploadDescription is not defined`

**Comment by jdlrobson** [comment]
Updating gist with comments.  ReferenceError is probably due to no requires statement in upload.js - it seems that previously this file has not been used and predates the use of require js.

File: assets/www/index.html (hunk 1 above)
Comment (yuvipanda) [inline assets/www/index.html]: I see whitespace after .desc :P

Comment (jdlrobson) [inline assets/www/index.html]: After the curly brace..

**Comment by awjrichards-zz** [comment]
Got the template working properly. Proof is in the pudding: http://test.wikipedia.org/wiki/File:Antiga_f%C3%A0brica_de_pells_(Escaldes-Engordany)_1342487002478.jpeg

File: assets/www/js/upload.js (hunk 5 above)
Comment (yuvipanda) [inline assets/www/js/upload.js]: Do we need to do that at all? I personally prefer not breaking them up.
